dpctl/tensor/_ctors.py

```diff
@@ -0,0 +1,377 @@
+            if usm_type not in ["device", "shared", "host"]:
+                raise ValueError(
+                    f"Unrecognized value of usm_type={usm_type}, "
+                    "expected 'device', 'shared', 'host', or None."
+                )
+        else:
+            raise TypeError(
+                f"Expected usm_type to be a str or None, got {type(usm_type)}"
+            )
+    # 5. Normalize device/sycl_queue [keep it None if was None]
+    if device is not None or sycl_queue is not None:
+        sycl_queue = _normalize_queue_device(q=sycl_queue, d=device)
+
+    # handle instance(obj, usm_ndarray)
+    if isinstance(obj, dpt.usm_ndarray):
+        return _asarray_from_usm_ndarray(
+            obj,
+            dtype=dtype,
+            copy=copy,
+            usm_type=usm_type,
+            sycl_queue=sycl_queue,
+            order=order,
+        )
+    elif hasattr(obj, "__sycl_usm_array_interface__"):
+        sua_iface = getattr(obj, "__sycl_usm_array_interface__")
+        membuf = dpm.as_usm_memory(obj)
+        ary = dpt.usm_ndarray(
+            sua_iface["shape"],
+            dtype=sua_iface["typestr"],
+            buffer=membuf,
+            strides=sua_iface.get("strides", None),
+        )
+        return _asarray_from_usm_ndarray(
+            ary,
+            dtype=dtype,
+            copy=copy,
+            usm_type=usm_type,
+            sycl_queue=sycl_queue,
+            order=order,
+        )
+    elif isinstance(obj, np.ndarray):
+        if copy is False:
+            raise ValueError(
+                "Converting numpy.ndarray to usm_ndarray requires a copy"
+            )
+        return _asarray_from_numpy_ndarray(
+            obj,
+            dtype=dtype,
+            usm_type=usm_type,
+            sycl_queue=sycl_queue,
+            order=order,
+        )
+    elif hasattr(obj, "__dlpack__"):
+        raise NotImplementedError(
+            "Support for __dlpack__ is not yet implemented"
+        )
+    elif isinstance(obj, (list, tuple, range)):
+        if copy is False:
+            raise ValueError(
+                "Converting Python sequence to usm_ndarray requires a copy"
+            )
+        _, dt, devs = _array_info_sequence(obj)
+        if devs == _host_set:
+            return _asarray_from_numpy_ndarray(
+                np.asarray(obj, dt, order=order),
+                dtype=dtype,
+                usm_type=usm_type,
+                sycl_queue=sycl_queue,
+                order=order,
+            )
+        # for sequences
+        raise NotImplementedError(
+            "Converting Python sequences is not implemented"
+        )
+    # obj is a scalar, create 0d array
+    return _asarray_from_numpy_ndarray(
+        np.asarray(obj),
+        dtype=dtype,
+        usm_type=usm_type,
+        sycl_queue=sycl_queue,
+        order="C",
+    )
+
+
+def empty(
+    sh, dtype="f8", order="C", device=None, usm_type="device", sycl_queue=None
```

Comment:
Parameter `usm_type="device"` by default for `empty`, but the parameter is `None` by default for `asarray`. Is it designed so?

Reply:
Yes. For `asarray( another_usm_array, usm_type=None)` will infer the `usm_type` from the input. If the input is not an object with `__sycl_usm_array_interface__`, the `usm_type` will be assume to be `"device"`.